// @augury/core/dist/plugins/popup-controller.class.js
export class PopupController {
    constructor(name) {
        this.name = name;
        this.storageKey = 'augury-popup-settings';
        this.onUnload = this.kill.bind(this);
        this.window = this.initializeWindow();
    }
    writeHtml(html) {
        this.window.document.open();
        this.window.document.write(html);
    }
    addScript(scriptContent) {
        const scriptElement = this.window.document.createElement('script');
        scriptElement.innerHTML = scriptContent;
        this.window.document.body.appendChild(scriptElement);
    }
    initializeWindow() {
        const parameters = ['titlebar=yes', 'location=no'].join(',');
        const pluginWindow = open('', this.name, parameters);
        if (!pluginWindow) {
            throw new Error('Please allow popup windows');
        }
        const sizeAndPosition = this.getWindowSizeAndPositionFromStorage();
        if (sizeAndPosition) {
            pluginWindow.moveTo(sizeAndPosition.position.x, sizeAndPosition.position.y);
            pluginWindow.resizeTo(sizeAndPosition.size.width, sizeAndPosition.size.height);
        }
        window.addEventListener('unload', this.onUnload);
        return pluginWindow;
    }
    getWindowSizeAndPositionFromStorage() {
        const fromStorage = localStorage.getItem(this.storageKey);
        return !!fromStorage ? JSON.parse(fromStorage) : null;
    }
    storeWindowSizeAndPosition() {
        // Fallback to screen width/height if popup is already closed
        const outerWidth = this.window.outerWidth === 0 ? screen.width : this.window.outerWidth;
        const outerHeight = this.window.outerHeight === 0 ? screen.height : this.window.outerHeight;
        const windowSizeAndPosition = {
            position: {
                x: this.window.screenX,
                y: this.window.screenY,
            },
            size: {
                width: outerWidth,
                height: outerHeight,
            },
        };
        localStorage.setItem(this.storageKey, JSON.stringify(windowSizeAndPosition));
    }
    kill() {
        this.storeWindowSizeAndPosition();
        this.window.close();
        window.removeEventListener('unload', this.onUnload);
    }
}
